src/lib/homestore_backend/hs_shard_manager.cpp

```diff
@@ -153,10 +153,26 @@ void HSHomeObject::on_shard_message_commit(int64_t lsn, sisl::blob const& header
     }
 
     // hs_ctx will be nullptr when HS is restarting and replay all commited log entries from the last checkpoint;
-    // but do we really need to handle this for create_shard or seal_shard?
-    // because this raft log had been commit before HO restarts and commit result is already saved in HS metablks
-    // when HS restarts, all PG/Shard infos will be recovered from HS metablks and if we commit again, it will cause
-    // duplication;
+    // most of time, the create_shard/seal_shard on_commit() is already completed before restarting/crash and shard info
+    // is already be written into metablk, so we need to do nothing in this case to avoid duplication when replay this
+    // journal log. but there is still a smaller chance that HO is stopped/crashed before writing metablk is called or
+    // completed and we need to recover shard info from journal log in such case.
+    sisl::sg_list value;
+    value.size = blkids.blk_count() * repl_dev->get_blk_size();
+    auto value_buf = iomanager.iobuf_alloc(512, value.size);
+    value.iovs.push_back(iovec{.iov_base = value_buf, .iov_len = value.size});
+    // header will be released when this function returns, but we still need the header when async_read() finished.
+    auto header_ptr = r_cast< const ReplicationMessageHeader* >(header.bytes);
+    repl_dev->async_read(blkids, value, value.size)
+        .thenValue([this, lsn, msg_header = *header_ptr, blkids, value](auto&& err) mutable {
+            if (err) {
+                LOGW("failed to read data from homestore pba, lsn:{}", lsn);
+            } else {
+                sisl::blob value_blob(r_cast< uint8_t* >(value.iovs[0].iov_base), value.size);
+                do_shard_message_commit(lsn, msg_header, blkids, value_blob, nullptr);
+            }
+            iomanager.iobuf_free(r_cast< uint8_t* >(value.iovs[0].iov_base));
+        });
```

Comment:
I don't understand the need for this addition. When the ReplDev restarts it will replay the operations it thinks are appropriate. In the future this will be driven by `nuraft::raft_server::start()`, but today it will be ReplSoloDev imitating that behavior. Given that, do we need to do anything special?

Reply:
based on the current behaviors of SoloReplDev, it will call ReplicaSetListener::on_commit() with hs_ctx been set to nullptr when replay the appropriate journal log. Even when we switch to real replication, ReplicaSetListener::on_commit()  will still will be called when replica restarts and replay the log and in that case hs_ctx will still be nullptr, right? 
we do not have the value buf which contains the shard info and has to read from the PBA.

Reply:
@hkadayam I think we've changed this correct? We shouldn't need to read from the Data service to commit a shard creation. The relevant data should be available somewhere in the commit context.

Reply:
I  revisit  the SoloReplDev restart logic and seems matches my expection here: https://github.com/eBay/HomeStore/blob/1fd188a87b07b776223fb009a59d2238197574b8/src/lib/replication/repl_dev/solo_repl_dev.cpp#L118.   on_commit() will be called with a PBA but leaving ctx to be nullptr.